ceph-mgr: Implement new pecan-based rest api

b-ranto opened the pull request:
The new rest API uses Flask with flask-restful extension and simplifies
the code significantly. It should be mostly equivalent in functionality
to the django-based rest API.

The new API uses Ceph auth system (requires caps mon allow *) and is
self-documenting via /doc endpoint.

Signed-off-by: Boris Ranto <branto@redhat.com>

jcsp commented:
Initial comments (sorry if this is all a bit critical, such is the nature of code review...)

When someone authenticates by CephX key, you're sending a remote command to the mon for every single HTTP request -- that's adding latency to the requests and adding load to the mons.  I imagine you're hoping that users will be well behaved and use the "token" mechanism most of the time -- I think that is quite optimistic.  IMHO if a user is going to require cephx credentials to initially create/destroy tokens, then that initial config might as well just be a CLI thing, and avoid the risk of users just using cephx keys for everything.

Encouraging people to send CephX tokens to an unauthenticated host is also just plain bad security.  While the connection is SSL, it is shipping with a self-signed certificate, so there is no authentication.

I can't see where the logic for modifying pg_num/pgp_num went (the part where we increase one first and then gradually increase the other).  One of the main reasons I wrote that for Calamari back in the day was an example of how API operations weren't always trivial things, and sometimes they had multiple steps -- the structure that enabled that should persist.

There doesn't seem to be any update to the ubuntu/debian packaging, are the dependencies present in Xenial?

I hope this doesn't seem too picky, but *please* can we avoid picking deliberately obscure names.  This module should probably just be called rest, or if we're going to have the old one co-existing, it should be called rest2 or rest-flask or something like that.

Exposing documentation on /doc is a neat, but exposing it on docs.ceph.com is much more important, is there a path to doing that?

If we're going to the trouble of having a brand new module, let's have some documentation to go with it, at least demonstrating how to authenticate.

jcsp commented on src/pybind/mgr/coleoid/common.py:
```diff
@@ -0,0 +1,176 @@
+    return " ".join(out)
+
+
+def HUMANIFY_REQUEST(request):
```
Any reason this isn't a method on CommandsRequest?

b-ranto replied:
It was on a TODO list, looks like I should have written it down. :)

jcsp commented on src/pybind/mgr/coleoid/common.py:
```diff
@@ -0,0 +1,176 @@
+
+
+# Transform command to a human readable form
+def HUMANIFY_COMMAND(command):
```
Function names should be lower_case

jcsp commented on src/pybind/mgr/coleoid/module.py:
```diff
@@ -0,0 +1,432 @@
+
+
+    @staticmethod
+    def run(commands, uuid=str(uuid4())):
```
This currently gets self.uuid passed in always, so no default needed, and instead of a static method that takes a uuid argument, this could be a class method that just uses self.uuid.

b-ranto replied:
True, it is a relict from one of the older ideas I had which I deprecated since then.

jcsp commented on src/pybind/mgr/coleoid/module.py:
```diff
@@ -0,0 +1,432 @@
+        try:
+            results = self.run(commands_arrays[0], self.uuid)
+        except:
+            instance.log.error(str(traceback.format_exc()))
```
This should mark itself failed at this stage, presumably.

b-ranto replied:
Looks like I forgot to remove this debugging try: except: statement which should no longer be necessary (I have previously had some issues with logging from different places so I had to catch this early but it is no longer necessary).

b-ranto commented:
> When someone authenticates by CephX key, you're sending a remote command to the mon for every single HTTP request -- that's adding latency to the requests and adding load to the mons. I imagine you're hoping that users will be well behaved and use the "token" mechanism most of the time -- I think that is quite optimistic. IMHO if a user is going to require cephx credentials to initially create/destroy tokens, then that initial config might as well just be a CLI thing, and avoid the risk of users just using cephx keys for everything.

Yep, I was hoping the users will use tokens when necessary. To create a persistent token all you need to do is to visit the /auth node while being authenticated by a CephX key. That will generate a persistent token (well, until you delete it with delete method or refresh it by visiting the /auth endpoint again).

You can access this easily via browser, since it sends back 401 with www-authenticate, the browser will prompt for the username/password. From python, all you need to do is add auth=('user', 'password') to the request method, i.e.
`requests.get(<url>, auth=('client.admin', <key>)`
or if you are using a token
`requests.get(<url>, auth=(<token>, '')`

btw: As far as speed goes, I have tested this in my VM with 1500 authenticated requests and it took 13.5 seconds when sending remote command for auth each time compared to about 10 seconds when using tokens.

> Encouraging people to send CephX tokens to an unauthenticated host is also just plain bad security. While the connection is SSL, it is shipping with a self-signed certificate, so there is no authentication.

I would not exactly call the host unauthenticated. It is running on a monitor and that is a host that is under the control of a Ceph admin. Also, paranoid admins can use their own keys for this and they can do the authentication properly. The packaging changes do not overwrite the existing keys on purpose. That being said, I admit I did not spend that much time on this part and I would be happy to do this better. I would e.g. like it if all the nodes did have the same cert/key, have some mechanism to verify the self-signed keys, etc.

btw: One of the reasons I wanted to use the CephX keys was that I was hoping we could automate the initial need for creation of ceph-mgr keys and we could use the key provided by a user to authenticate ceph-mgr at a later point and only then, it would start serving requests. Although I did not look into how easy/difficult that would be, yet.

> I can't see where the logic for modifying pg_num/pgp_num went (the part where we increase one first and then gradually increase the other). One of the main reasons I wrote that for Calamari back in the day was an example of how API operations weren't always trivial things, and sometimes they had multiple steps -- the structure that enabled that should persist.

The submit_request method accepts list of lists. The pg_num is changed in the first iteration and only if that succeeds the command will change the pgp_num. I may have missed something there though. i.e. the request looks like

`[[change pg_num, change stg, change stg_else], [change pgp_num]]`

The request mechanism will run the commands in the first sub-list in parallel and once those are done it will run the commands in the second sublist (well, the one command). If you are saying gradually then maybe there should be more sub-lists?

> There doesn't seem to be any update to the ubuntu/debian packaging, are the dependencies present in Xenial?

Sigh, I always forget about these. The ssl bits will definitely be there. I have just checked for the python-flask-restful extension on pkgs.org and it seems to be available in ubuntu xenial as well as debian sid.

> I hope this doesn't seem too picky, but please can we avoid picking deliberately obscure names. This module should probably just be called rest, or if we're going to have the old one co-existing, it should be called rest2 or rest-flask or something like that.

Sure, the primary reason I went for an obscure name was that 'rest' was already taken and I did not want to do a PR that would patch the old rest directory as that would look like a total mess on github. It might be ok if I did that in two commits I suppose (one that removed the old rest api and one that added the flask-based one).

btw: My working name was neverest, I can switch to that if you like it better?

> Exposing documentation on /doc is a neat, but exposing it on docs.ceph.com is much more important, is there a path to doing that?
>
> If we're going to the trouble of having a brand new module, let's have some documentation to go with it, at least demonstrating how to authenticate.

I do plan to create a (html-formatted) snapshot of the /doc endpoint once the API stabilizes and document it on docs.ceph.com. No automated path for that, yet. :)

jcsp commented:
> btw: As far as speed goes, I have tested this in my VM with 1500 authenticated requests and it took 13.5 seconds when sending remote command for auth each time compared to about 10 seconds when using tokens.

I'm not so concerned about the API performance as I am about it placing extra load on the mons (and also spamming the audit log).  If someone is running e.g. a UI that's doing several requests every few seconds, it's unreasonable to be sending a command to mon for every one of those.

I really wouldn't rely on users to be well-behaved enough to switch auth methods to the more efficient tokens.  If you just avoid giving them the cephx option to begin with, we don't have to worry.

Users/tools would have to go the CLI or their keyring file to get their cephx token to begin with, so it's probably not at all painful to just have them use the CLI to create their token for API access.

> I would not exactly call the host unauthenticated. It is running on a monitor and that is a host that is under the control of a Ceph admin.

From the REST client's point of view, it is just talking to an IP address and hoping the thing responding is really the monitor.  This is a "red address bar" situation, where we have a SSL connection but we are only hoping that the thing sending response packets is really the host we thought we were talking to.  The host definitely is unauthenticated in the security sense of the word, there's no ambiguity about that.

> Also, paranoid admins can use their own keys for this and they can do the authentication properly.

At the risk of sounding a bit preachy, we have a duty to ship things as secure as we can by default, not rely on admins to secure them after the fact.

Self-signed certificates are an unfortunate reality in some environments, but we can mitigate this by avoiding sending over-privileged secrets (like cephx keys) over that channel.  I think there's also an argument that we should really wave this issue in the admin's face by requiring them to either load their own SSL key, or at least requiring them to manually type a command that tells them they are creating a self-signed certificate.

> The request mechanism will run the commands in the first sub-list in parallel and once those are done it will run the commands in the second sublist (well, the one command). If you are saying gradually then maybe there should be more sub-lists?

Look at PgCreatingRequest.  It needs to batch the creation of PGs and wait for them to be done before creating some more.

That reminds me of a more general point -- the commands in the current rest code (OsdMapModifyingRequest etc) will make sure the up-to-date cluster maps are loaded before considering the job complete.  That's so that if someone does a POST and then a GET, the results reflect the change they just made.  It could be reasonable to just not try and do that any more, but it is a change you are making that you should be aware of.

> btw: My working name was neverest, I can switch to that if you like it better?

That's better but my preference would be for something that starts with "rest".  See how ceph-devel feels about just replacing the old one outright.  You need to make some effort to find out about anyone who is using the old one, to have the discussion about whether to kill it.

> I do plan to create a (html-formatted) snapshot of the /doc endpoint once the API stabilizes and document it on docs.ceph.com. No automated path for that, yet. :)

Ah, good.  Hopefully that will include some text explaining things, and an example piece of client code etc, not just a list of fields.

b-ranto commented:
Latest changelog:
- renamed to restful
- the SSL key is generated at runtime (first node that starts generates it), not in packaging
- the SSL key/cert is stored as a ceph config-key and shared between the nodes
- increased the permissions for mon profile mgr (we need to be able to write config-keys and read(+x) auth)
- implemented a shutdown method+endpoint (for internal use only, requires a master key that is not shared anywhere and is unique to each running instance) to do a clean shutdown
- minor fixes
- deployed the latest chacra build with ansible, passed my testing

b-ranto commented:
**security**: We do not authenticate the servers in the sense of SSL but afaik, we don't do that with Ceph tools either (*). The SSL certificates just make it more obvious because they are designed to be signed by a CA but they do have a use even when they are just self-signed -- you can easily detect a change in host (certificate changed), the communication is encrypted,...

It is only the first encounter with the server when you are vulnerable to the MITM attack just like you are with Ceph itself (imagine someone spoofing your DNS records and acting as a middle man resending hte data to the real Ceph node, the person would have your data, know your key, ...).

In any case, the new way of handling the certificates allows you to verify the certificate via the usual channels (`ceph config-key get mgr.restful.cert`) and compare it to the key reported by the REST api server running on port 8002. This gives you the same level of server authentication as the cli commands have. This could also probably be further automated -- we could e.g. provide a CLI utility that checks that the two certificates are the same.

There is also the point that if you use a non-standard auth system, the admins are very likely to not take the API that seriously but that is really rather bad -- the API is rather powerful (you can even delete pools with it...). If you use the CephX key for auth, it makes this much more obvious and it should make users much more cautious about it which is a good thing.

(*)  We would need a third party -- some kind of CA -- for that. I really can't imagine how a server authentication without some kind of CA would work and afaik, we do not have a CA in Ceph.

**speed**: My point was that 20-30% performance penalty if you are running the API as a user (i.e. one request per few seconds or minutes) does not sound that bad. On the other hand, if you are doing several requests per second with a web UI, it is really rather easy to generate a token and use the token-based authentication.

jcsp commented:
> security: We do not authenticate the servers in the sense of SSL but afaik, we don't do that with Ceph tools either (*)

Without having double checked this, I believe that CephX authentication doesn't transmit the key.  It's shared secret crypto: the server proves to the client that it already knows the client's key, and the client proves to the server that it has the key, all without actually sending the key.

The central issue with what you're doing here is that you're transmitting that key like a password, but it's not a password.  It's a key for use in shared secret crypto.

> speed: My point was that 20-30% performance penalty if you are running the API as a user (i.e. one request per few seconds or minutes) does not sound that bad. 

I don't care about the penalty for the rest api consumer, I care about it spamming the monitors.  I don't want every monitor log we see to be a long stream of "auth get" commands, and I don't want monitors to have to service these extra requests unnecessarily.

> On the other hand, if you are doing several requests per second with a web UI, it is really rather easy to generate a token and use the token-based authentication.

Certainly it's easy to generate a token, but that doesn't mean users will do it.  There's no need to offer them the choice -- just let the tokens be configured via the CLI, and then only allow HTTP access with tokens.

jcsp commented on src/pybind/mgr/restful/module.py:
```diff
@@ -0,0 +1,511 @@
+    def _shutdown(self):
+        # We need request module to get the shutdown function
+        requests.post(
+            'https://localhost:8002/shutdown',
```
Does the werkzeug server really not have a better way to shut down than making a process POST to itself?

b-ranto replied:
Werkzeug does but Flask (that runs the werkzeug server) does not provide it directly so you need the request to get the environment variable that points to the shutdown function. Some more details can be found e.g. here:

http://stackoverflow.com/questions/15562446/how-to-stop-flask-application-without-using-ctrl-c

jcsp commented on src/mon/MonCap.cc:
```diff
@@ -173,7 +173,8 @@ void MonCapGrant::expand_profile_mon(const EntityName& name) const
     profile_grants.push_back(MonCapGrant("osd", MON_CAP_R | MON_CAP_W));
-    profile_grants.push_back(MonCapGrant("config-key", MON_CAP_R));
+    profile_grants.push_back(MonCapGrant("auth", MON_CAP_R | MON_CAP_X));
+    profile_grants.push_back(MonCapGrant("config-key", MON_CAP_R | MON_CAP_W));
```
Hmm, so it looks like @liewegas rolled back the permissions of the mgr quite a bit, in practice I suspect we're going to end up adding back read/write access to most of these in order to enable mgr modules to do management stuff.

b-ranto replied:
Actually, when I first tested this, even the "osd" had only MON_CAP_R so I could not e.g. create a pool. I think @tchaikov changed that in the last couple of days.

jcsp commented:
@tchaikov @dmick you guys may also have some comments on this?

tchaikov commented:
@jcsp will review this change (and yours) in one day or two.

liewegas commented:
that's fine!

dmick commented:
I've been trying to find any rationale or discussion, and failing.  Is there some background for "why are we doing this"?

b-ranto commented:
@dmick Dependencies, the django based REST api  won't run unless you manually (with pip I believe) install all the dependencies. This one only requires stuff that is already in EPEL. It was easier for me to do a complete rewrite based on the django based REST api than trying to strip the django dependencies from the current API. This runs out of the box, no manual steps are necessary and we can package it easily.

jcsp commented:
(Aside: I am still surprised that a total rewrite was believed to be less effort than generating a Django RPM)

Speaking of rationales, I notice that this still hasn't been brought up on ceph-users, ceph-devel, or at a CDM call.  If the hope here is that this API will be used by anybody, then it would probably make a lot of sense to try and find some would-be users and/or see if anyone was depending on the old rest module and/or the old calamari api.

In fairness to @b-ranto, he initially just pointed me at his branch, and I asked for a PR so that we had a place to leave comments/discussion.  So now that we've got a PR, where do we want to go from here?  Is this something that somebody wants to include in luminous?  Are there tests on the way[1]?

1. the rest module did not have tests in the tree, but it did have the old calamari tests that were ready to be ported over mostly as-is

b-ranto commented:
Latest changelog:
- switched to pecan due to the demand by other developers in the team and availability (no new non-rhel dependencies)
- switched the authentication to the one in the old rest api but it does not work because the handle_command interface is broken, probably since commit 23575077 (it always says command not supported although ceph tell can find the commands)

b-ranto commented on src/pybind/mgr/restful/api.py:
```diff
@@ -0,0 +1,659 @@
+        # Check for invalid pool args
+        invalid = common.invalid_pool_args(args)
+        if invalid:
+            return {'message': 'Invalid arguments found: "%s"' % str(invalid)}, 500
```
I was thinking about that but for the sake of consistency I used 500 almost everywhere (apart from auth where I used 401 to prompt for login/password). I suppose the important thing here is it is not 2xx.

b-ranto commented:
@jcsp Packaging Django is highly non-trivial and a long-term project but it is not just the dependencies, there are several people in our team that are familiar with pecan and it is generally easier to maintain thanks to it being much more lightweight (the same holds for flask).

The primary user of the API should be storage console/tendrl which wants to switch from calamari to ceph-mgr and it is already familiar with pecan-based rest APIs (e.g. ceph-installer).

ktdreyer commented:
> we have a duty to ship things as secure as we can by default

Agreed, and this includes keeping up with Django CVEs :) It's not just about generating a Django RPM (or set of RPMs), but staying on top of all the vulnerabilities that stack would entail.

liewegas commented:
It seems like this is going to be the path forward.  I think we need to have *some* API in place (see e.g., #14778).  Assuming for a moment that the answer is "yes, we want this," what needs to be done before we can merge it?

b-ranto commented:
There are currently some issues with the handle_command interface in master so the username/password creation via cli does not work. Other than that I would be fine removing DNM. I can revert the patch that changed the auth system back to CephX which would resolve the issue since it did not require the handle_command interface.

b-ranto commented:
Rebased and retested, the hande_command interface was fixed in the meantime, removing DNM.

tserong commented:
Just to muddy the waters... [openATTIC](https://www.openattic.org/) is another management tool that (without speaking for the openATTIC team) will want to talk to mgr. At the moment, openATTIC has its own internal REST API for talking to the Ceph cluster, which was initially developed quite some time ago (it uses librados internally).  openATTIC deliberately didn't use the Calamari API, partly due to uncertainty about the future of Calamari at the time, and also because openATTIC is already a Django app, but apparently the Calamari API didn't use Django models and was thus somehow difficult to work with.

openATTIC's REST API endpoints are listed at https://wiki.openattic.org/display/OP/openATTIC+Ceph+REST+API+overview. I had a notion to try to lift the openATTIC REST API piece out of the openATTIC source tree and into mgr, and see what happens  (although I
haven't actually done this yet).  This would give mgr yet another REST API...

Does this Pecan API yet have any users?  Is there scope to discuss this with the Tendrl folks, the openATTIC folks, anyone else who wants an API and figure out what everyone needs?  Or am I about to blunder into https://xkcd.com/927/?

tchaikov commented:
@tserong i will take a look at our rest_api.py tests to see if we can switch it to your RESTful implementation.

liewegas commented:
As with the calamari API, we probably want an API call that just passes 
through the "cli" mon/mgr commands.  I suspect it's as simple as changing 
the URL endpoint and making sure the new API impl has a similar 
pass-through call.

dmick commented:
FWIW, I tried to run this branch last night with the system packages for pecan on Xenial, and failed:

2017-05-04 00:44:38.235274 7fd59bd46700  0 mgr[restful] Traceback (most recent call last):
  File "/home/dmick/src/ceph/src/pybind/mgr/restful/module.py", line 221, in serve
    self._serve()
  File "/home/dmick/src/ceph/src/pybind/mgr/restful/module.py", line 259, in _serve
    ssl_context=context
  File "/usr/lib/python2.7/dist-packages/werkzeug/serving.py", line 512, in make_server
    passthrough_errors, ssl_context)
  File "/usr/lib/python2.7/dist-packages/werkzeug/serving.py", line 450, in __init__
    self.socket = ssl_context.wrap_socket(self.socket,
AttributeError: 'Context' object has no attribute 'wrap_socket'

I've not yet started to dig.

b-ranto commented:
@dmick It looks like, the later versions of werkzeug no longer support OpenSSL library, I will have to switch to ssl which will hopefully be supported everywhere.

dmick commented:
Ah, yes, as of 0.10 it looks like

dmick commented:
https://github.com/dmick/ceph/commit/f89561848e60301205d2209e25ee8a0ff7988889 works, but it's not very appetizing

b-ranto commented:
@dmick I was doing pretty much the same in my local repo but with NamedTemporaryFile which cleans itself on closure. I did not publish it because I was working on a couple more things (splitting the api file, fixing GET input method which was broken, support for 'wait' argument on requests, etc).

b-ranto commented:
I decided to get rid of the pyOpenSSL dependency and reworked the cert/key system to allow your own certificates by writing them to a fixed location in /etc/ceph/ with the latest commit 6bf168c. The self-signed certificates are still shared amongst the mgr nodes.

b-ranto commented:
The latest changelog/comments:
- the code was cleaned up and modularized (REST api was turned into a module, minor fixes, etc)
- implemented a simple pass through method, you just need to POST to /request endpoint with a proper python structure describing a command in the POST body
- implemented ?wait=1 http get argument that will make an async request wait for its completion
- improved key/cert handling, fixed compatibility with werkzeug 0.10+
- rebased and patched for new mgr API differences (send_command changed type)
- stripped pyOpenSSL dependency

FWIW I would be quite happy if the REST API got merged at this point.

@liewegas a simple pass through method was implemented in the API

@tserong I am in favour of merging the various REST APIs into a single one, most of the OpenATTIC endpoints should be covered by this REST API, if you point me to the sources, I can take a look/guide you at implementing the rest of the endpoints

tserong commented:
@b-ranto, the openATTIC endpoints are:

````
api/ceph/
api/ceph/<fsid>
api/ceph/<fsid>/osds[?name=<name>][?ordering=...]
api/ceph/<fsid>/osds/<id>
api/ceph/<fsid>/pools[?name=<name>][?ordering=...]
api/ceph/<fsid>/pools/<id>
api/ceph/<fsid>/pgs[?osd_id=<osd-id>][?pool_name=<pool-name>][?ordering=...]
api/ceph/<fsid>/pgs/<id>
api/ceph/<fsid>/erasure-code-profiles[?name=<name>][?ordering=...]
api/ceph/<fsid>/erasure-code-profiles/<id>
api/ceph/<fsid>/rbds[?name=<name>][?ordering=...]
api/ceph/<fsid>/rbds/<id>
api/ceph/<fsid>/fs
````
This is set up by https://bitbucket.org/sebastian-wagner/openattic/src/cb4c39a92c7516164f8ef82144c302842d58854a/backend/ceph/urls.py?at=OP-2055-oa-classic-split&fileviewer=file-view-default

The models backing it are in https://bitbucket.org/sebastian-wagner/openattic/src/cb4c39a92c7516164f8ef82144c302842d58854a/backend/ceph/models.py?at=OP-2055-oa-classic-split&fileviewer=file-view-default

(Apologies for brevity, I'll be looking at this properly tomorrow)

jcsp commented on src/pybind/mgr/restful/module.py:
```diff
@@ -0,0 +1,501 @@
+        return request
+
+
+    def run_command(self, command):
```
This function appears to be unused

b-ranto replied:
You are right. It was used by the cephx-based authentication to get info about the key right away but that is no longer the case so it is unused.

jcsp commented:
I can't find where the discussion happened around flask vs. pecan etc.  

Has there been some discussion on the list?  What is the feedback from Tendrl and OpenAttic developers?

Where are the tests?

What's the plan for versioning the API endpoints?

I'm not sure adding that CLI pass through is such a brilliant idea.  The experience so far has shown up problems with that:
* people think that because they're going through the REST API that the commands they're sending are a stable/supported thing
* people get confused about the level of functionality of the REST API, because less-technical people hear "service X is doing Y via the REST API", when in reality they're doing it via the CLI and it's just passing through a REST interface.
* people who can hack something together using the CLI pass through are not well motivated to contribute to improving the REST interface to properly wrap the functionality they care about.

I say that as the person who added the CLI pass through to Calamari.  We've lived through that experiment and it was a real pain, let's not repeat it.

b-ranto commented:
> I can't find where the discussion happened around flask vs. pecan etc.
Mostly in person/calls.
> What's the plan for versioning the API endpoints?
There is the api_version field in the / (root) endpoint.
> Where are the tests?
I'm currently testing it with a ~simple script but that requires a running cluster at the moment. I can look into turning into some basic sanity tests.

As far the pass through method, it is really trying to be as stupid as possible:
- the doc directly states it is a pass through method
- you need to post to /request endpoint and you need to post the dict with 'prefix', etc for it to work so that should make it even more clear
- it returns the request data with outs, outb, etc directly (+ a couple of useful dict arguments)

That being said, I don't have any strong feelings about that one way or the other.

liewegas commented:
We had a CDM discussion about this last week.  @tserong is verifying that OpenAttic to migrate to this without much pain (I think mostly we'll need to add the rbd stuff, but the rest is cosmetic?).

As far as API stability, I would even be happy if we merged this and called it "experimental" for the luminous cycle as gaps and shortcomings are identified and fixed.  I suspect the api versioning can make this a reasonably non-painful thing, but I'm not the expert here.  My feeling is that something, even if it is a somewhat unstable moving target, is better than nothing.

Retaining the mon cli API passthrough was my suggestion.  Is there a way to indicate it's not necessarily a stable API?

tserong commented:
I'm doing some random testing/experimentation (thought it'd be interesting to compare this and openATTIC on the same cluster) and have found two kinks so far.

The first is that it can't create an SSL cert, because the unprivileged ceph user can't write to the /etc/ceph/ directory (it's 0755, owned by root). This manifests as the following in /var/log/ceph/ceph-mgr.$id.log:

````
mgr[restful] Traceback (most recent call last):
  File "/usr/lib64/ceph/mgr/restful/module.py", line 215, in serve
    self._serve()
  File "/usr/lib64/ceph/mgr/restful/module.py", line 235, in _serve
    ssl_context=self.load_cert(),
  File "/usr/lib64/ceph/mgr/restful/module.py", line 344, in load_cert
    make_ssl_devcert(cert_base, host='localhost')
  File "/usr/lib/python2.7/site-packages/werkzeug/serving.py", line 327, in make
_ssl_devcert
    with open(cert_file, 'w') as f:
IOError: [Errno 13] Permission denied: '/etc/ceph/ceph-mgr-restful.crt'
````

I worked around that with `chown ceph /etc/ceph`, but that's probably not ideal.

The second kink is that if I make an HTTP request, the connection just gets reset:

````
# curl http://localhost:8002/api/
curl: (56) Recv failure: Connection reset by peer
````

There doesn't appear to be anything useful in the log in this case. Not sure what's up with that yet.

b-ranto commented:
@tserong it is SSL so you need to use https, not http

I thought the /etc/ceph should be owned by ceph user for quite some time, now (and in fact in my cluster installation it is). I can look into that further.

tserong commented:
> @tserong it is SSL so you need to use https, not http

Ugh. I are idiot. Too long at the screen, apparently. https works fine...

> I thought the /etc/ceph should be owned by ceph user for quite some time, now (and in fact in my cluster installation it is). I can look into that further.

Not as shipped in the Ceph RPM -- https://github.com/ceph/ceph/blob/master/ceph.spec.in#L1058 doesn't specify a user, so it picks up %defattr(-,root,root,-).

b-ranto commented:
> Not as shipped in the Ceph RPM -- https://github.com/ceph/ceph/blob/master/ceph.spec.in#L1058 doesn't specify a user, so it picks up %defattr(-,root,root,-).

Yeah, I already noticed that, I wonder what made it (ceph, ceph) on my cluster (afaicr, I did not do that manually). I am currently looking at storing it in some other location.

tserong commented:
/var/lib/ceph/mgr/$cluser-$id/... ?

b-ranto commented:
Yep, there is really rather no other place to store it, I have the patch in the works.

b-ranto commented:
Should be fixed with the latest commit. I used mgr_data (`/var/lib/ceph/mgr/$cluster-$id`) to store the keyring file with fallback to /tmp.

liewegas commented:
What about storing it in mon config-key?

If the user is going ot use a non-self-signed cert, that is probably where 
it should go.

tserong commented:
Can we use a single SSL cert for all mgr instances, or do we need a separate cert per mgr instance so that the certificate subject will match the system hostname?

liewegas commented:
What do API endpoints for other services (e.g., openstack stuff) do for this sort of thing?

b-ranto commented:
The way it currently works is that the self-signed certificates are stored as config-keys (cert, key) so that they could be shared between the mgr instances (we are talking the contents of the files, not their location). If you override the key by creating a file for the cert it won't be shared making the user responsible for sharing the keys. I added this option so that it is more user-friendly to override the key/cert pairs.

I need to store the keys somewhere because ssl_context expects to be given file names, not the contents of the files. The full logic is defined here (the function makes sure the cert/key files exist and returns their file names as a pair):

https://github.com/ceph/ceph/blob/wip-rest-flask/src/pybind/mgr/restful/module.py#L320

I am currently thinking that maybe it would be best if I added a ceph tell mgr command that would store a user-supplied SSL key/cert pair as config-keys and store the files in the system only as temporary files.

b-ranto commented:
> Can we use a single SSL cert for all mgr instances, or do we need a separate cert per mgr instance so that the certificate subject will match the system hostname?

We kinda have to use a single certificate for all the nodes if we want to be able to route the traffic to the active mgr instance from non-active ones which is certainly something I would like to see. Otherwise, the cert would change ~randomly.

jcsp commented:
Using a different certificate per node is the only way to have a green address bar, unless someone constructs wildcard ssl certificates and makes sure all their mgr instances are communicating from a *.ceph.mysite.com domain.

However, in the case where someone is using a "red address bar" approach (crude self signed cert with no authority loaded into the client web browser), it's probably annoying to force them to load N separate such certs if all their servers are really going to use the same one.

It would be nice to have non-active mgr instances respond with redirects to the active mgr, I don't think having distinct certificates presents any obstacle to that?  What kind of routing were you thinking of that would need a shared certificate?

b-ranto commented:
@jcsp that is pretty much how it works now, you can give it your own certificate and that one won't be shared, if you don't give it any, it will create a self-signed certificate and share it amongst the hosts

By 'kinda have to' I pretty much meant that it would be a pain to work with a server that changes its certificate a lot (well, if you want to verify its authenticity at least to a certain degree). By that I mean that the shared self-signed key would at least allow you to verify that server you are talking to has not changed (well, the group of servers you are talking to).

Anyway, I am currently thinking it might be best if we did not try to be too smart and if we just generated the cert/key pair in ceph-mgr post script (no sharing). That way, it could be stored in /etc/ceph/ and would be cleaned on removal.

liewegas commented:
Okay, can/should we store the cert in config-key instead of /etc/ceph?

sebastian-philipp commented:
Hi all,

let me review the API from an openATTIC perspective. I won't go into any implementation details here, instead I focus solely on the API Design.

This is mainly a whishlist of items, including questions or notes.

TL;DR: I really like the API, but I can see some rough edges at the moment.

**Endpoint `/config/cluster`**

I really would like to have real JSON types, instead of just strings. For example, instead of

    bluestore_max_alloc_size: "0",

I'd love to have 

    bluestore_max_alloc_size: 0,

Also, which are the default values for each setting?

The output is not really a RESTful collection. Instead, what about returning an array of JSON objects?

**Endpoint `/mon`**

As we don't export MONs at the moment, I cannot give you any feedback here.

**Endpoint `/osd`**

We're showing the current usage of OSDs: How big is it and how much free space is left. Is ther any chance of adding this?

Do you support pagination here?

**Endpoint `/osd/<arg>/command`**

How am I supposed to trigger a command via PATCH?

**Endpoint `/pool`**

This looks like an implementation detail that should not be exposed:

    read_tier: -1,
    tier_of: -1,

Instead, I'd omit them or set them to `null`.

Is there a documentation of which properties can be PATCHED? E.g. do you support PATCHing `read_tier` and `tier_of`?

Again, we need the current usage statistics for pools, including a percent value, if possile.

Please change `type: 1` to `type: "replicated"`, There is no documentation of the pool types, except for a C++ header somewhere in the sources.

flags_names should be a list of flag names. E.g.  `flags_names: ["hashpspool", "xyz"]` instead of `flags_names: "hashpspool, xyz"`

Regarding POSTing new pools. Either change `pool_name` to `name` in GET responses, or change `name` to `pool_name` in POST requests.

**Endpoint `/request`**

To be RESTful, `/request` should be an array of objects instead of an object. Why are you not returning all data in `/request`, similar to all other endpoints?

The `running` property is really confusing. What is the meaing of this?

    "running": [
        [
            "osd pool create pg_num=8 pool=test",
            "",
            ""
        ]
    ],


POSTing to `/request` sounds strage. Instead, I would expect to get a request url back after increasing the pg_num. 

In openATTIC, we're requests provide a progress in percent. This is especially important when creating lots of pgs. As a user, I would really appriciate an estimation, when a request will be finished.


**pagination**

Do you have a concept of using pagination? Listing all PGs won't work without pagination at all.

**Filter**

Is there a concept of filters? I'd like to filter OSDs by server. E.g. by calling `/osd/?server=my-server`. Also, we're supporting filtering OSDs and pools by name. 

**Missing Endpoints**

In order to use this API, we're going to need `/erasure-code-profile`, `/rbd` and `/pg`


**Error Handling**

I'd not return Python Tracebacks for user errors. Expect that we will forward any error to our users.


    $ echo <<EOF | curl -d @- -s 'https://127.0.0.1:8002/pool' --insecure
      {
       "pg_num": 8,
       "pool_name": "test",
      }
    EOF
    {
        "message": [
            "Traceback (most recent call last):",
            "  File \"/home/sebastian/Repos/ceph/src/pybind/mgr/restful/decorators.py\", line 42, in decorated",
            "    return f(*args, **kwargs)",
            "  File \"/home/sebastian/Repos/ceph/src/pybind/mgr/restful/decorators.py\", line 33, in decorated",
            "    return f(*args, **kwargs)",
            "  File \"/home/sebastian/Repos/ceph/src/pybind/mgr/restful/api/pool.py\", line 105, in post",
            "    args = request.json",
            "  File \"/usr/local/lib/python2.7/dist-packages/pecan/core.py\", line 77, in __getattr__",
            "    return getattr(obj, attr)",
            "  File \"/usr/local/lib/python2.7/dist-packages/pecan/core.py\", line 52, in __getattribute__",
            "    return WebObRequest.__getattribute__(self, name)",
            "  File \"/usr/local/lib/python2.7/dist-packages/webob/request.py\", line 724, in _json_body__get",
            "    return json.loads(self.body.decode(self.charset))",
            "  File \"/usr/lib/python2.7/json/__init__.py\", line 339, in loads",
            "    return _default_decoder.decode(s)",
            "  File \"/usr/lib/python2.7/json/decoder.py\", line 364, in decode",
            "    obj, end = self.raw_decode(s, idx=_w(s, 0).end())",
            "  File \"/usr/lib/python2.7/json/decoder.py\", line 382, in raw_decode",
            "    raise ValueError(\"No JSON object could be decoded\")",
            "ValueError: No JSON object could be decoded",
            ""
        ]
    }

To sum up, I'm really looking forward to use this API in the future.

alfredodeza commented on src/pybind/mgr/restful/decorators.py:
```diff
@@ -0,0 +1,56 @@
+
+
+# Helper function to catch and log the exceptions
+def catch(f):
```
this can be done with a Pecan hook, that way you don't need to decorate every method. There is a 'on_error' hook that can be used for this.

My comment shouldn't prevent this from moving forward though, this is OK for now :)

b-ranto replied:
I switched to the on_error hook with the latest commit. The catch function had the benefit of returning the traceback directly to the user, not just to the ceph-mgr log but that was mostly a convenience hack for me while I was testing this.

jcsp commented on debian/ceph-mgr.postinst:
```diff
@@ -24,6 +24,13 @@ set -e
+	PKEY="/etc/ceph/ceph-mgr-restful.key"
+	if [ ! -e "$CERT" -o ! -e "$PKEY" ]; then
+	    openssl req -new -nodes -x509 \
+		-subj "/C=US/ST=Oregon/L=Portland/O=IT/CN=ceph-mgr-restful" \
```
Can we leave the US/Oregon/Portland parts out?

b-ranto replied:
I was not sure whether those are mandatory or not but my testing suggests we can left them out.

jcsp commented on src/pybind/mgr/restful/module.py:
```diff
@@ -227,12 +227,17 @@ def _serve(self):
         )
 
+        ssl_context = (
+            '/etc/ceph/ceph-mgr-restful.crt',
```
This path should be something configurable so that we can do things like make vstart.sh work, and also perhaps so that people can run two mgr instances on the same node with different certs.

The convenient place from a runtime point of view would be in ``mgr_data``, but of course at package installation time that location is unknown.

What if the module looked for certs in both locations, so that anyone who wanted a per-daemon cert could put it in the /var/lib... location, and otherwise would get the global one generated at installation time?

b-ranto replied:
It is now configurable per node via config-keys. The fallback is in /etc/ceph/.

b-ranto commented:
I have switched the port from 8002 to 8003. The reason is that it allows us to run (and compare) both calamari and ceph-mgr. It also gives a hint on how to do backwards-incompatible changes (i.e. api versioning that breaks things). We can simply bump the port and have both daemons running at the same time (if necessary). The api_version field in root endpoint can serve as an indicator on how advanced the API is at the moment, i.e. we should bump it every time we introduce a new endpoint.

GregMeno commented:
@b-ranto We've got python-pecan-1.0.2, python-werkzeug-0.9.1, and python-flask-0.10.1
Are those new enough?
Do we need other stuff?

@jcsp 
How do those packages get installed in general? Are they added as deps to the ceph-mgr package.

b-ranto commented:
@GregMeno I am testing this on rhel 7, the versions I get there are python-pecan-0.4.5-2.el7.noarch and python-werkzeug-0.9.3-1.el7_0.noarch. I no longer need any flask bits. Both dependencies are currently part of the ceph-mgr package both in deb and rpm.

b-ranto commented:
@sebastian-philipp thanks for the review, I will try to explain/comment to most of the points you mentioned, here

`/config/cluster`

Here, I just json-serialize the config that I get. It is what ceph(-mgr) exports as a config. Unfortunately, it does not seem to support any types at the moment. While we could still try to add some native types (i.e. detect if it is int and convert to int if it is), this might cause more problems than it fixes. Imagine an option that can be int-like (but does not have to). The option would sometimes be converted to int but not always and that would be a pain to work with.

`/osd`

This could probably use some work on it. It would be nice to give additional information. As we have access to pretty much all mon and osd commands, it should not be too difficult to add the additional stats.

`/osd/<arg>/command`

PATCH is just another request method. Most utilities should already support it (it has been around for quite some time). However, if your utility/library does not support it, pecan allows you to do a POST request to `/osd/<arg>/command?_method=patch` (the same holds e.g. for DELETE).

btw: Judging from [this](http://stackoverflow.com/questions/16306624/how-to-use-patch-verb-with-curl), curl seems to have some issues with patch if you use -X instead of --request.

`/pool`

This sounds like this could use some work. I'll need to look into it later.

`/request`

The running property is the list of running (sub-)commands for a given request. The first line is the command that is running, the second and third are the outputs. The request can run several commands in parallel (or sequentially) as requested. A request is a collection of `running` commands, `finished` commands and commands `waiting` to be executed. 

POST on request gives a simple pass-through method to run any command, even those that are not supported by the API. A regular API endpoint that runs command will do this for you and return request itself. Maybe I could add an 'url' key to the dict that is returned but it is just `https://<url>:8003/request/<id>`.

I suppose I can add the percentage as the number of already finished commands (compared to the total number of commands being run). Do you return an actual percentage (i.e. max 100) or number between 0  and 1? I think I would prefer the second method.

`filters`

There already are some filters but they are not widely used. It would certainly be nice to improve on this. We can consider this an RFE. :-)

`error handling`

This was changed by the latest commit. I had it this way mostly for convenience when debugging (I did not have to go through logs to see what went wrong).

`pagination`

To be honest, I am not exactly sure what this is supposed to mean/do so I can't really comment. :-/

tserong commented:
I think one of the things @sebastian-philipp was getting at with requests was that he'd like the ability to get some sort of request URL or ID back after initiating a long running operation triggered by a POST/PATCH to one of the other API endpoints. For example, if you use the /osd/<arg>/... endpoints to initiate a scrub or reweight, or even mark an OSD down/out, there needs to be a means of knowing when that operation is complete. Likewise creating a pool or changing the pg_num using the (future) /pool/... endpoints. This is probably further complicated in the case where, say, the mgr you're talking to goes down part way through a long-running operation - can the mgr that takes over still tell the client about the status of those long running tasks?

Pagination is the ability to split data across several pages. Consider a cluster with hundreds of OSDs, or thousands of PGs - rather than returning every single one of those objects in response to /osd or /pg, instead it returns a "page" of, say, the first 100 objects. Then you can get the next page by appending page=2 to the URL, and so forth.

b-ranto commented:
@tserong When an endpoint does a request, you get the full request back, including its id that you can poll, its current state (that will be running), the commands it scheduled and so on. Unfortunately, the actual completion percentage is missing at the moment. When you poll the id for the completion you get the same request back although updated. You can even make the API wait for the request to complete with ?wait=1, then you always get a finished request even if it is in a failed state. It looks like, the /request endpoint deserves more documentation.

Oh, that pagination. Yeah, that is not supported at the moment but it should not be too difficult to implement. I can imagine having a decorator that will do the pagination for us. The /request endpoint would actually be any of those that could benefit from pagination. I will try to come up with something ~soon.

tserong commented:
@b-ranto cool, thanks for the clarification :-)

alfredodeza commented:
@b-ranto python-pecan-0.4.5-2.el7.noarch is way too old, I think you will at least want to try with what the ceph-installer ships with (1.0.2 I think).

See: https://pecan.readthedocs.io/en/latest/changes.html

b-ranto commented:
@alfredodeza as long as it is backwards-compatible, it should not cause much trouble. Anyway, I have retested with pecan-1.1.2 and it seems to work fine as well (it passed my test script).

b-ranto commented:
@sebastian-philipp @tserong I have added a paginate decorator with the latest commit so there already is a pagination support. Please note that it is per request (you need to give it ?page=N, otherwise it will return the full list).

liewegas commented:
@b-ranto Looks like this needs a rebase to fix the ceph.spec.in conflict.  Also, what does your test script look like for this?  Can we include that as a workunit and/or a teuthology test? 

I suggest we merge this more or less as is and then address any changes to existing endpoints that we need to/can before luminous.  The librbd support will be somewhat involved, but useful not just to the openattic folks!

b-ranto commented:
@liewegas I have rebased and fixed the conflicts. I use this simple script to test the code:

https://pastebin.com/AnYgJPXZ

It mostly just checks the code integrity by going through almost all the endpoints and checking that they return http 200. It could use a bit more automation e.g. to pass the admin key, etc.

liewegas commented:
Cool, can you add that script to qa/workunits/rest/test_mgr_rest_api.py (or something like that), and make it take the endpoint via the command line?  Thanks!

b-ranto commented:
I guess you meant the admin key should be a cli argument. I have added the test script.

b-ranto commented:
Can we merge this?

liewegas commented:
Oops!  Yeah, I thought we already had.

jacob-go commented:
Hello everyone,

I hope I'm not rude for commenting after a while, but I'm trying to work with this module right now and have some problems. 

Actually, my only problem is with the `/request` endpoint. I don't understand the structure I need to use to get this endpoint responding. You talked about "a proper python structure" and "prefix", but unfortunately it's not clear for everybody. 

The command I'm issuing right now looks something like:
`curl -k https://<SERVER_IP>:8003/request -X POST -H "Authorization: Basic <BASE64_ENCODED_STRING>" -d '{"ceph": "status"}'`. So as you can see the structure I'm using is just a plain JSON, validated by lint. It's not working also with `{"ceph": status}`. 

The logs say thus:

> 2017-08-17 16:10:07.009483 7fba5bd89700  0 mgr[restful] Traceback (most recent call last):
>   File "/usr/lib/python2.7/dist-packages/pecan/core.py", line 678, in __call__
>     self.invoke_controller(controller, args, kwargs, state)
>   File "/usr/lib/python2.7/dist-packages/pecan/core.py", line 572, in invoke_controller
>     result = controller(*args, **kwargs)
>   File "/usr/lib/ceph/mgr/restful/decorators.py", line 33, in decorated
>     return f(*args, **kwargs)
>   File "/usr/lib/ceph/mgr/restful/api/request.py", line 87, in post
>     return module.instance.submit_request([[request.json]], **kwargs)
>   File "/usr/lib/ceph/mgr/restful/module.py", line 585, in submit_request
>     request = CommandsRequest(_request)
>   File "/usr/lib/ceph/mgr/restful/module.py", line 66, in __init__
>     results = self.run(commands_arrays[0])
>   File "/usr/lib/ceph/mgr/restful/module.py", line 84, in run
>     result.command = common.humanify_command(commands[index])
>   File "/usr/lib/ceph/mgr/restful/common.py", line 38, in humanify_command
>     out = [command['prefix']]
> KeyError: 'prefix'
> 2017-08-17 16:10:07.009483 7fba5bd89700  0 mgr[restful] Traceback (most recent call last):
>   File "/usr/lib/python2.7/dist-packages/pecan/core.py", line 678, in __call__
>     self.invoke_controller(controller, args, kwargs, state)
>   File "/usr/lib/python2.7/dist-packages/pecan/core.py", line 572, in invoke_controller
>     result = controller(*args, **kwargs)
>   File "/usr/lib/ceph/mgr/restful/decorators.py", line 33, in decorated
>     return f(*args, **kwargs)
>   File "/usr/lib/ceph/mgr/restful/api/request.py", line 87, in post
>     return module.instance.submit_request([[request.json]], **kwargs)
>   File "/usr/lib/ceph/mgr/restful/module.py", line 585, in submit_request
>     request = CommandsRequest(_request)
>   File "/usr/lib/ceph/mgr/restful/module.py", line 66, in __init__
>     results = self.run(commands_arrays[0])
>   File "/usr/lib/ceph/mgr/restful/module.py", line 84, in run
>     result.command = common.humanify_command(commands[index])
>   File "/usr/lib/ceph/mgr/restful/common.py", line 38, in humanify_command
>     out = [command['prefix']]
> KeyError: 'prefix'

I must admit I don't understand the prefix thing. And maybe I'm wrong in the understanding of the endpoint itself as a whole - shouldn't I use it to pass ANY ceph command? It'll be great if somebody would explain. Maybe the developer himself @b-ranto?

b-ranto commented:
Hi,

that method is a simple pass through method for arbitrary commands as defined in `src/mon/MonCommands.h`. Let's consider the following command
```
COMMAND("osd ls " \
        "name=epoch,type=CephInt,range=0,req=false", \
        "show all OSD ids", "osd", "r", "cli,rest")
```
The prefix is `osd ls` (or `osd ls `, not exactly sure which one is the correct one at the moment, even both of them might work). The optional argument's name is `epoch` and it is of type `CephInt` (python's int should work just fine).

I.e. You should send something like `{'prefix': 'osd ls', 'epoch': 0}` to the endpoint for it to work.

The endpoint is hard to use by design. It is not recommended to post to the endpoint unless you know exactly what you are doing. There is no parsing of output, no error handling, nothing. It will just pass the python structure and return the output.

jacob-go commented:
Thank you SO much @b-ranto ! 
I don't know if it's your original intention, but that explanation really helped me, so I could quickly post the command and query the result.

b-ranto commented:
Yeah, feel free to play with it or use it, just really don't expect anything special out of it. We had some issues in the past with a similar pass through method that was easier to use where people expected it to do way more than it actually did.

jacob-go commented:
@b-ranto No, be certain I'm not after anything special, just run basic commands via REST. :)

Liuchang0812 commented on src/pybind/mgr/restful/module.py:
```diff
@@ -0,0 +1,487 @@
+
+            else:
+                self.keys[command['key_name']] = str(uuid4())
+                self.set_config_json('keys', self.keys)
```
hi, a quick question: `self.set_config_json` probably run failed, like bad authorization. should we process this case here?

b-ranto replied:
This code has changed a lot since this PR has been merged. Feel free to create an upstream issue or PR if you think there is something wrong with the current code.

However, in general, set_config_json is unlikely to fail. If it does it may point to some deeper issues with ceph-mgr code.

Liuchang0812 replied:
thanks @b-ranto roger that.

jacob-go commented:
@b-ranto I guess the answer is 'no', but I have to ask just in case you know better - Is it possible to invoke the Ceph admin socket (under `/var/run/ceph/`) via the `/request` endpoint? Execute something like `ceph --admin-daemon $file perfcounters_dump` or `ceph daemon osd.0 perf dump`? I guess it's problematic since it's outside the scope of `MonCommands`, and maybe a bit dangerous in general.

jcsp commented:
If this discussion could move into a ceph-devel mailing list thread for broader visibility (and so that I don't keep getting notifications from this long-closed PR?), I would appreciate it
